rfcs/rfc-111.md

```diff
@@ -0,0 +1,354 @@
+    Properties.BlockDeviceMappings.*.device_name == /^\/dev\/sdc-\d/ # all other local must have sdc
+}   
+```
+
+### Enhanced outputs with named rules
+
+Because rules can be explicitly named in rulesets, we can also provide this to customers in the rule evaluation output, with an explicit PASS/FAIL/SKIP (When a "when" block is not satisfied) next to each rule's name. Evaluation will output an overall status of a ruleset against a given template, along with evaluation status of all named rules and their dependent clauses as well. This gives customers a large amount of visibility into which exact rule/clause failed for easier debugging. Console output would be of the form:
+
+```
+Ruleset RULESET_FILE_NAME: Overall Status PASSED
+/named_rule_one: PASSED
+/named_rule_one/AWS::S3::Bucket/Clause#linenumber: PASSED
+/AWS::S3::Bucket/Clause#linenumber: FAILED  (additional info on failures)
+```
+
+With such a structured format for output, this will allow the tool to handle different output formats like JSON in addition to the console output for easier parsing of rule evaluation results.
+
+## Drawbacks/Open Questions
+
+As mentioned above, this approach would change the root at which type clauses are evaluated. Instead of the properties section of a given resource type, the properties available for evaluation are from the object containing properties, e.g.
+
+```json
+{
+    "Type": "AWS::S3::Bucket",
+    "Propeties":  {
+        "BucketName": "MyBucket"
+    },
+    "DeletionPolicy": "Retain"
+}
+```
+
+instead of
+
+```json
+{
+    "BucketName": "MyBucket"
+}
+```
+
+the change would require checks on resource properties to be prefixed with "Properties." to maintain parity. Requesting more feedback on what users prefer. The change makes certain operations like accessing deletion policy less ambiguous and does away with the [relative operator](https://github.com/aws-cloudformation/cloudformation-guard/tree/master/cfn-guard#checking-resource-properties-and-attributes) used to check deletion policy and resource attributes. We feel this makes the language less ambiguous in an exchange for more explicitly specifying propety paths.
+
+An alternative considered is to allow for nested blocks inside of type blocks to allow for easier checking, so a block like:
+```
+AWS::EC2::Instance {
+    Properties.BlockDeviceMappings.*.Ebs != null
+    Properties.BlockDeviceMappings.*.device_name == /^\/dev\/ebs-/
+    Properties.BlockDeviceMappings.*.Ebs.encrypted == true
+    Properties.BlockDeviceMappings.*.Ebs.delete_on_termination == true
+}
+```
+
+is equivalent to:
+```
+AWS::EC2::Instance {                      
+    Properties {
+        BlockDeviceMappings.*.Ebs != null
+        BlockDeviceMappings.*.device_name == /^\/dev\/ebs-/
+        BlockDeviceMappings.*.Ebs.encrypted == true
+        BlockDeviceMappings.*.Ebs.delete_on_termination == true
+    }
+}
+```
+
+Would like to see the preference of users on this specific point as to what is easiest to use.
```

Comment:
Personally would prefer not having to specify the top-level `Properties` in almost every rule

Reply:
Why not using the [spring boot yaml format](https://docs.spring.io/spring-boot/docs/current/reference/html/spring-boot-features.html#mapping-yaml-to-properties) or sort of, with indentation for each sub-level. The `Properties` would be visible once, and same for each sub-level. Last sample could be simplified like this:

AWS::EC2::Instance {                      
    Properties {
        BlockDeviceMappings {
           *.Ebs != null
           *.device_name == /^\/dev\/ebs-/
           *.Ebs.encrypted == true
           *.Ebs.delete_on_termination == true
        }
    }
}
Not sure however if the * could be used for a block...

Reply:
I like this way of specifying a lot to reduce repetition. Let me see how we can incorporate this into the block evaluation.